src/wp-includes/block-template-utils.php

```diff
@@ -1,23 +1,481 @@
+		),
+		'privacy-policy' => array(
+			'title'       => __( 'Privacy Policy' ),
+			'description' => '',
+		),
+		'404'            => array(
+			'title'       => _x( '404', 'Template name' ),
+			'description' => __( 'Template shown when no content is found.' ),
+		),
+	);
+
+	/**
+	 * Filters the list of template types.
+	 *
+	 * @param array $default_template_types An array of template types, formatted as [ slug => [ title, description ] ].
+	 *
+	 * @since 5.x.x
+	 */
+	return apply_filters( 'default_template_types', $default_template_types );
+}
+
+/**
+ * Checks whether the input 'area' is a supported value.
+ * Returns the input if supported, otherwise returns the 'uncategorized' value.
+ *
+ * @access private
+ * @since 5.9.0
+ *
+ * @param string $type Template part area name.
+ *
+ * @return string Input if supported, else the uncategorized value.
+ */
+function _filter_block_template_part_area( $type ) {
+	$allowed_areas = array_map(
+		function ( $item ) {
+			return $item['area'];
+		},
+		get_allowed_block_template_part_areas()
+	);
+	if ( in_array( $type, $allowed_areas, true ) ) {
+		return $type;
+	}
+
+	/* translators: %1$s: Template area type, %2$s: the uncategorized template area value. */
+	$warning_message = sprintf( __( '"%1$s" is not a supported wp_template_part area value and has been added as "%2$s".' ), $type, WP_TEMPLATE_PART_AREA_UNCATEGORIZED );
+	trigger_error( $warning_message, E_USER_NOTICE );
+	return WP_TEMPLATE_PART_AREA_UNCATEGORIZED;
+}
+
+/**
+ * Finds all nested template part file paths in a theme's directory.
+ *
+ * @access private
+ * @since 5.9.0
+ *
+ * @param string $base_directory The theme's file path.
+ * @return array $path_list A list of paths to all template part files.
+ */
+function _get_block_templates_paths( $base_directory ) {
+	$path_list = array();
+	if ( file_exists( $base_directory ) ) {
+		$nested_files      = new RecursiveIteratorIterator( new RecursiveDirectoryIterator( $base_directory ) );
+		$nested_html_files = new RegexIterator( $nested_files, '/^.+\.html$/i', RecursiveRegexIterator::GET_MATCH );
+		foreach ( $nested_html_files as $path => $file ) {
+			$path_list[] = $path;
+		}
+	}
+	return $path_list;
+}
+
+/**
+ * Retrieves the template file from the theme for a given slug.
+ *
+ * @access private
+ * @since 5.9.0
+ *
+ * @param string $template_type wp_template or wp_template_part.
+ * @param string $slug template slug.
+ *
+ * @return array|null Template.
+ */
+function _get_block_template_file( $template_type, $slug ) {
+	$template_base_paths = array(
+		'wp_template'      => 'block-templates',
+		'wp_template_part' => 'block-template-parts',
+	);
+	$themes              = array(
+		get_stylesheet() => get_stylesheet_directory(),
+		get_template()   => get_template_directory(),
+	);
+	foreach ( $themes as $theme_slug => $theme_dir ) {
+		$file_path = $theme_dir . '/' . $template_base_paths[ $template_type ] . '/' . $slug . '.html';
```

Comment:
What happens if the `$template_type` is not in the `$template_base_paths`? Will the value passed to it _always_ be `wp_template` or `wp_template_part`? If yes, then it's fine. Else, consider a guard clause at the start of the function such as:

```php
if ( 'wp_template' !== $template_type || 'wp_template_part' !== $template_type ) {
	return null;
}
```

Reply:
Will defer to @ockham here 🙂

Reply:
I think bailing early makes sense, since we don't totally control what template type is being passed here.

Fixed in f4d281c. (I changed the `||` to an `&&` though, since I think that's the right condition here.)

Reply:
GB counterpart PR: https://github.com/WordPress/gutenberg/pull/36318